packages/patterns/src/components/rename-block-modal-control.js

```diff
@@ -0,0 +1,176 @@
+
+	const closeModal = () => setRenamingBlock( null );
+	const handleSubmit = () => {
+		const message =
+			nameIsOriginal || nameIsEmpty
+				? sprintf(
+						/* translators: %s: new name/label for the block */
+						__( 'Block name reset to: "%s".' ),
+						editedBlockName
+				  )
+				: sprintf(
+						/* translators: %s: new name/label for the block */
+						__( 'Block name changed to: "%s".' ),
+						editedBlockName
+				  );
+
+		// Must be assertive to immediately announce change.
+		speak( message, 'assertive' );
+
+		// If the new value is the block's original name (e.g. `Group`)
+		// or it is an empty string then assume the intent is to reset
+		// the value. Therefore reset the metadata.
+		let newName = editedBlockName;
+		if ( newName === blockInformation?.title || ! newName.trim() ) {
+			newName = undefined;
+		}
+
+		updateBlockAttributes( [ clientId ], {
+			metadata: {
+				...metadata,
+				name: newName,
+			},
+		} );
+
+		// Immediate close avoids ability to hit save multiple times.
+		closeModal();
+	};
+
+	return (
+		<Modal
+			title={ __( 'Rename' ) }
+			onRequestClose={ closeModal }
+			overlayClassName="block-editor-block-rename-modal"
+			focusOnMount="firstContentElement"
+			aria={ { describedby: descriptionId } }
+			size="small"
+		>
+			<form
+				onSubmit={ ( e ) => {
+					e.preventDefault();
+
+					if ( ! isNameValid ) {
+						return;
+					}
+
+					handleSubmit();
+				} }
+			>
+				<VStack spacing="3">
+					<Text id={ descriptionId }>
+						{ __( 'Enter a custom name for this block.' ) }
+					</Text>
+
+					<TextControl
+						__nextHasNoMarginBottom
+						__next40pxDefaultSize
+						value={ editedBlockName }
+						label={ __( 'Block name' ) }
+						hideLabelFromVision
+						help={
+							hasPatternOverrides
+								? __(
+										'This block allows overrides. Changing the name can cause problems with content entered into instances of this pattern.'
+								  )
+								: undefined
+						}
+						placeholder={ blockInformation?.title }
+						onChange={ setEditedBlockName }
+						onFocus={ autoSelectInputText }
+					/>
+					<HStack justify="right">
+						<Button
+							__next40pxDefaultSize
+							variant="tertiary"
+							onClick={ closeModal }
+						>
+							{ __( 'Cancel' ) }
+						</Button>
+
+						<Button
+							__next40pxDefaultSize
+							aria-disabled={ ! isNameValid }
+							variant="primary"
+							type="submit"
+						>
+							{ __( 'Save' ) }
+						</Button>
+					</HStack>
+				</VStack>
+			</form>
+		</Modal>
+	);
+}
+
+// Split into a different component to minimize the store subscriptions.
+export default function RenameBlockModalControl() {
+	const { renamingBlockClientId } = useSelect(
+		( select ) => ( {
+			renamingBlockClientId: unlock(
+				select( patternsStore )
+			).getRenamingBlockClientId(),
+		} ),
+		[]
+	);
+	if ( ! renamingBlockClientId ) return null;
+
+	return <BlockRenameModal clientId={ renamingBlockClientId } />;
+}
```

Comment:
Wondering if you'd considered the `BlockSettingsMenuControls` slot to add the pattern specific rename action? It looks like the fills handily get rendered right after the rename menu item:
https://github.com/WordPress/gutenberg/blob/8deb8f92a30973c47a3336114893b477f008ffa3/packages/block-editor/src/components/block-settings-menu-controls/index.js#L86-L91

Then it might just require a `allowBlockRenaming: false` editor setting then to remove the normal rename action.

Reply:
My idea is to avoid customizing through the controls so that we can reuse the same `renameBlock` function for commands and other renaming flows as well. We can consider supporting both if needed though, but to me it's not necessary for now. Does it make sene?